Use precompiled binaries when available.

cocoa-xu opened the pull request:
Precompiled binaries will be used by default if they are available. This is mostly useful for the following reasons:
- when the toolchain is not available (e.g, running livebook on some embedded devices, or running in nerves environment, where a C compiler is not shipped)
- a working C compiler will not be a strict requirement.
- save time on compiling.

However, this behaviour can be disabled by setting the `STB_IMAGE_PREFER_PRECOMPILED` environment variable to `NO`.

The `ARCH-OS-ABI` is detected in the following order:
1. `ARCH-OS-ABI` is set in the environment variable `STB_IMAGE_HOST`. By setting this environment variable, you can force the use of a specific precompiled binary.
2. `TARGET_ARCH`, `TARGET_OS`, `TARGET_ABI` are all set in the environment.
3. When `uname` is available, it will be used to determine the architecture and OS.

Precompiled binaries will  be available for the following platforms when a tag is pushed (which triggers the precompile workflow for Linux and macOS):
- x86_64-apple-darwin
- arm64-apple-darwin
- x86_64-linux-gnu
- aarch64-linux-gnu
- arm-linux-gnueabihf
- riscv64-linux-gnu
- s390x-linux-gnu
- ppc64le-linux-gnu

cocoa-xu committed:
precompile workflow for linux/macos

cocoa-xu committed:
Install Erlang and Elixir

cocoa-xu committed:
only zip what's inside priv

cocoa-xu committed:
cxx_flags => cpp_flags

cocoa-xu committed:
use precompiled binaries if allowed and possible

cocoa-xu committed:
include precompiled_deploy.exs in package files

josevalim commented:
Great work @cocoa-xu!

Unfortunately this is a bit more complicated because we need to precompile per NIF version and ideally we also want to store checksums. You can see how RustlerPrecompiled tackle those specific problems: https://github.com/philss/rustler_precompiled

I wonder if we could somehow reuse rustler precompiled or zig to handle those. Or I guess we prepare a precompilation toolchain for C code, like rustlerprecompiled.

cocoa-xu commented:
Yeah, I can do that. Probably the only difference is that there will be a workflow file for each system, Linux, macOS and maybe Windows in the future. Because Cross-compiling to different systems is hard for C/C++. 

Finding appropriate and well-maintained prebuilt C/C++ toolchains for this task is possible but definitely not as easy as `rustup target add "ARCH-VENDOR-SYSTEM-ABI"`. Therefore, using GitHub action seems to be a better approach (since it provides native Linux/macOS/Windows environments and C/C++ toolchains).

If this is okay, I'll refactor the code. :)

cocoa-xu commented:
I guess I need to add that this also means if users want to compile from source, they can only get binaries for the host OS (including different arches and ABIs for the same OS). 

For example, compiling from source on Linux will generate binaries for  the following targets (by default):

- x86_64-linux-gnu
- aarch64-linux-gnu
- arm-linux-gnueabihf
- riscv64-linux-gnu
- s390x-linux-gnu
- ppc64le-linux-gnu

While compiling on macOS will only generate binaries for these targets:
- x86_64-apple-darwin
- arm64-apple-darwin

It's possible to compile binaries for Linux on macOS using cross-compile toolchains or even using docker, but compiling binaries on Linux for macOS is really not as easy.

josevalim commented:
I believe the Zig compiler can cross compile C files and, given this project is not really complex, it can be a suitable option here?

cocoa-xu committed:
refactored

cocoa-xu committed:
fix vendor and cleanup

cocoa-xu committed:
fix compile args

cocoa-xu committed:
use mac for precompiling

cocoa-xu committed:
include checksum in release

cocoa-xu committed:
include checksum in release

cocoa-xu committed:
updated fennec_precompile

cocoa-xu committed:
updated fennec_precompile

cocoa-xu committed:
updated fennec_precompile

cocoa-xu committed:
updated to latest fennec_precompile

cocoa-xu committed:
update workflow

cocoa-xu committed:
update mix.lock

cocoa-xu committed:
use hex fennec_precompile

cocoa-xu commented:
> I believe the Zig compiler can cross compile C files and, given this project is not really complex, it can be a suitable option here?

Okay, I got it working. This should be much simpler now. :)

josevalim commented:
Absolutely beautiful! Some additional questions:

1. Do we need zig only for precompilation or also for regular compilation?
2. Can you please provide a guide similar to https://hexdocs.pm/rustler_precompiled/precompilation_guide.html that tells us how to precompile and release?

josevalim commented:
Some additional information:

1. Generally speaking, you don't want to hardcode on Mix: https://github.com/cocoa-xu/fennec_precompile/blob/main/lib/fennec_precompile.ex#L7 - It is fine at compile time, but it is best to pass the otp_app as argument, as Rustler precompiled does

2. With RustlerPrecompiled, we have a force_build that defaults to true if the version is `-dev`. This is important in development so we can build things without precompiling. You can also set `force_build` via a configuration, which is important for companies with more strict security requirements that don't  want to rely on precompiled binary. Here are the docs for RustlerPrecompiled: https://github.com/philss/rustler_precompiled/blob/main/lib/rustler_precompiled.ex#L31-L37 - we need to support similar functionality and it would be great if, when forcing build, we don't need Zig installed. :)

But this is great!! Sorry for the delay on reviewing!

josevalim commented:
Hi @cocoa-xu, I have been thinking about this idea throughout the idea, and I also discussed with @whatyouhide, and I think we can simplify this.

You don't actually need `use Fennec.Precompile` module. The reason why `Rustler.Precompile` does it is because Rustler itself uses a module to compile stuff, but I don't think we should.

So here is my suggestion:

1. Change `compilers: [:elixir_make] ++ Mix.compilers()` to `compilers: [:fennec_precompile] ++ Mix.compilers()`

2. From the `:fennec_precompile` task, you can access the version, the app name, and everything without worries. If you need new configuration parameters, you can add them to `def project` in `mix.exs`. Such as `fennec_nif_filename` and `fennec_download_url`

3. The new fennec_precompile task will attempt to download the precompiled artifacts and put them in priv unless in `force_build` mode. As per the previous comment, we will force build whenever the dependency has a -dev "pre" flag or an app configuration was explicitly set. In force_build mode, we will build the artifacts without the zig toolchain (we can have an option to enable the toolchain though)

4. The fennec.precompile, fennec.fetch, and checksum stuff will continue as is.

This should make it closer in spirit to `elixir_make`. Then we can test it out and, if this works out, then I propose we merge it into `elixir_make` and make only the precompilation toolchain bit configurable.

What do you think?

I am copying @fhunleth because he is likely interested in this. :)

cocoa-xu committed:
update use-clause

cocoa-xu commented:
Yeah, I think I've implemented these points in the newer version of `fennec_precompile`. Just pushed it to the repo.

I think there are only some minor differences between the ideas here and the current implementation ([de9f90b](https://github.com/cocoa-xu/fennec_precompile/tree/de9f90b98ece62530e112285b6c6866522a67705)) of `fennec_precompile`. Basically, the question/difference is when should the compilation happen if `force_build` is `true`.

Currently, the answer is: 
1. It can be deferred to the very last minute, when the module that contains the `use`-clause is being compiled, if `:fennec_precompile` was only declared in the `deps` and `force_build` is `true`.
    ```elixir
    defp deps do
        [
            {:fennec_precompile, "~> 0.1"},
            # ....
        ]
    end
    ```

2. It can also happen when the library user tries to compile the project if `:fennec_precompile` is used as a compiler.
    ```elixir
    def project do
        [
            compilers: [:fennec_precompile] ++ Mix.compilers()
            # ...
        ]
    end

    defp deps do
        [
            {:fennec_precompile, "~> 0.1"},
            # ...
        ]
    end
    ```

If `force_build` is `false`, then the download happens when the module that contains the `use`-clause is being compiled.

cocoa-xu commented:
Okay, I refactored `:fennec_precompile`, it should run as what is proposed here.🥳 And I also made a tiny example [fennec_example](https://github.com/cocoa-xu/fennec_example).

cocoa-xu committed:
compile with fennec_precompile

cocoa-xu committed:
fix ci

cocoa-xu committed:
bump fennec_precompile to v0.1.3

josevalim commented on .github/workflows/macos.yml:
```diff
@@ -18,17 +18,25 @@ jobs:
+          import Config
+
+          config :fennec_precompile, :config, stb_image: [
+            fennec_force_build: true,
```
Instead of setting `force_build` on the CI file, you can change the version in the mix.exs to `-dev`. :)

josevalim commented:
This is awesome @cocoa-xu! Now my next question is how we can abstract this and bring into elixir_make. I think `elixir_make` can be responsible for:

1. building checksums and unpacking tar.gz's into priv
2. downloading files and verifying the checksum
3. Own the new tasks elixir_make.precompile and elixir_make.fetch

The precompiler is a module that needs to have three functions:

1. Return the current target
2. Return a list of supported target
4. Call precompile with the given config + targets and it returns a list of {target, tar_gz}

My suggestion is for us to design the `@behaviour` for the precompiled module within `fennec`, and then move it around to `elixir_make` once we are all happy. :)

cocoa-xu commented:
Hi @josevalim, thanks for the suggestions! I've refactored `fennec` accordingly. 

[`lib/*.ex`](https://github.com/cocoa-xu/fennec_precompile/tree/main/lib) now only contains all the necessary functions for the `FennecPrecompile.Precompiler` behaviour and some helper functions for identifying target triplets. 

[`lib/mix/tasks/*.ex`](https://github.com/cocoa-xu/fennec_precompile/tree/main/lib/mix/tasks) are the three mix tasks, `compile`, `precompile` and `fetch`. They should be probably ok to move to `elixir_make` with some minor changes.

After moving mix tasks to `elixir_make`, `fennec_precompile` will only define the `FennecPrecompile.Precompiler` behaviour along with some helper functions. [branch - behaviour-only](https://github.com/cocoa-xu/fennec_precompile/tree/behaviour-only/lib)

josevalim commented:
Amazing! Can you please send a PR for `elixir_make`? Then we can iron out the final details there. :)

Once we move it there, enabling precompilation should be a matter of doing two steps:

1. Add `fennec_precompiler` as a dependency
2. Adding `make_precompiler: FennecPrecompiler` to the mix.exs

Thank you! :)

cocoa-xu commented:
Hi @josevalim, I've opened a PR for `elixir_make`, that code is ready for review! I'll add more inline docs later. :)

cocoa-xu committed:
resolve conflicts

cocoa-xu committed:
Merge branch 'main' into cx-precompile

cocoa-xu committed:
precompile for x86_64-windows-msvc

cocoa-xu committed:
updated deps

cocoa-xu committed:
mix format

cocoa-xu committed:
updated deps

cocoa-xu committed:
manually compile the project before mix test

cocoa-xu committed:
manually precompile the project before mix test

cocoa-xu commented:
Hi @josevalim, I updated the code to match the incoming changes in elixir_make. Also, I've added the `x86_64-windows-msvc` target to the precompile workflow. 

At the moment,  I'm using `github` links for the `elixir_make` and `cc_precompiler` in `deps`, I'll update them later once they are merged and available in hex.pm.

cocoa-xu committed:
use otp 25.0.4. fix TLS :client: In state :hello_middlebox_assert at ssl_gen_statem.erl:736

cocoa-xu committed:
update deps

cocoa-xu committed:
ci-precompile-linux: install more cross-compilers

cocoa-xu committed:
Merge branch 'main' into cx-precompile

cocoa-xu committed:
update deps

cocoa-xu committed:
ci: bump otp and elixir version

cocoa-xu committed:
ci: remove `_build` directory after precompilation

cocoa-xu committed:
ignore `checksum.exs` file

cocoa-xu committed:
include `checksum.exs` in hex package

cocoa-xu committed:
updated target checks

cocoa-xu committed:
updated deps

josevalim commented:
@cocoa-xu it is finally time! Can you please update this PR to use elixir_make v0.7.0? Also, could you please write a small RELEASE.md file with the steps to release a new version? It can be something as simple as:

```
1. git tag NEW_VERSION
2. Wait for workflows
3. mix elixir_make.checksum
4. mix hex.publish
```

I just need to know the overall steps. :)

Thank you! :heart:

cocoa-xu commented:
Sure! I'll update and publish the `cc_precompiler` and then push new commits here. :)

cocoa-xu committed:
using cc_precompiler v0.1.0

cocoa-xu committed:
added RELEASE.md

cocoa-xu committed:
Merge branch 'main' into cx-precompile

cocoa-xu committed:
updated `project/0`

cocoa-xu commented:
CI is all green! Ready for review now  :) @josevalim

cocoa-xu committed:
updated README.md

josevalim commented on RELEASE.md:
```diff
@@ -0,0 +1,8 @@
+
+The overall steps to release `StbImage` with its precompilation binaries are:
+
+1. git tag NEW_VERSION
+2. Wait for workflows
+3. mix elixir_make.checksum
+4. mix hex.publish
```
```suggestion
1. `git tag NEW_VERSION`
2. `git push --tags` and wait for workflows to run
3. `mix elixir_make.checksum`
4. `mix hex.publish`
```

josevalim commented on README.md:
````diff
@@ -16,6 +16,19 @@ def deps do
 end
 ```
 
+This library also provides some precompiled binaries for the following platforms:
````
```suggestion
This library also provides some precompiled binaries for the following platforms:

```

josevalim committed:
Apply suggestions from code review

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

josevalim commented:
Awesome work! I just shipped v0.6.0 and it worked like a charm!!!

cocoa-xu commented:
> Awesome work! I just shipped v0.6.0 and it worked like a charm!!!

Thanks @josevalim! I was quite nervous, but luckily it seems that nothing goes wrong. Hooray!
